src/scheduler/FunctionCallClient.cpp

```diff
@@ -96,24 +93,62 @@ void clearMockRequests()
 }
 
 // -----------------------------------
-// gRPC client
+// Message Client
 // -----------------------------------
 FunctionCallClient::FunctionCallClient(const std::string& hostIn)
-  : host(hostIn)
-  , channel(grpc::CreateChannel(host + ":" + std::to_string(FUNCTION_CALL_PORT),
-                                grpc::InsecureChannelCredentials()))
-  , stub(faabric::FunctionRPCService::NewStub(channel))
-{}
+  : faabric::transport::MessageEndpointClient(hostIn, FUNCTION_CALL_PORT)
+{
+    this->open(faabric::transport::getGlobalMessageContext(),
+               faabric::transport::SocketType::PUSH,
+               false);
+}
+
+FunctionCallClient::~FunctionCallClient()
+{
+    if (!faabric::util::isMockMode()) {
+        this->close();
+    }
+}
+
+void FunctionCallClient::sendHeader(faabric::scheduler::FunctionCalls call)
+{
+    // Deliberately using heap allocation, so that ZeroMQ can use zero-copy
+    int functionNum = static_cast<int>(call);
+    size_t headerSize = sizeof(faabric::scheduler::FunctionCalls);
+    char* header = new char[headerSize];
+    memcpy(header, &functionNum, headerSize);
+    // Mark that we are sending more messages
+    send(header, headerSize, true);
+}
+
+void FunctionCallClient::awaitResponse(char*& data, int& size)
+{
+    // Call the superclass implementation
+    MessageEndpointClient::awaitResponse(
+      faabric::util::getSystemConfig().endpointHost,
+      FUNCTION_CALL_PORT + REPLY_PORT_OFFSET,
+      data,
+      size);
+}
 
 void FunctionCallClient::sendFlush()
 {
-    faabric::Message call;
+    faabric::ResponseRequest call;
     if (faabric::util::isMockMode()) {
         flushCalls.emplace_back(host, call);
     } else {
-        ClientContext context;
-        faabric::FunctionStatusResponse response;
-        CHECK_RPC("function_flush", stub->Flush(&context, call, &response));
+        // Send the header first
+        sendHeader(faabric::scheduler::FunctionCalls::Flush);
+
+        // Send the message body
+        call.set_returnhost(faabric::util::getSystemConfig().endpointHost);
+        size_t msgSize = call.ByteSizeLong();
+        char* serialisedMsg = new char[msgSize];
+        // Serialise using protobuf
+        if (!call.SerializeToArray(serialisedMsg, msgSize)) {
+            throw std::runtime_error("Error serialising message");
+        }
+        send(serialisedMsg, msgSize);
```

Comment:
Yes, the macro is looking great. Wrt the `delete` I deliberately omitted it. This is because I am setting the cleanup callback in 0MQ's send, for it to be zero copy. Note that this was taken from a [2010 tutorial](http://wiki.zeromq.org/blog:zero-copy). More on this on the other comments.